### PR TITLE
Refactor IFederationResolver; add FederationResolverAttribute

### PR DESCRIPTION
**Pull request by Shane32**
#3921 Federation step 4: Add federation resolver attribute

## Summary

My goal here is to support two different ways to write federation resolvers in type-first schemas:

1. Via static method, where the method arguments are the various representation properties.  When there is a single `Id` field used to represent an entity, this probably makes the most sense from a CLR perspective.

```cs
// configure federation key fields
[Key("id")]
public class Widget
{
    // configure fields
    [Id]
    public string Id { get; set; }
    public string Name { get; set; }

    // configure federation resolver
    [FederationResolver]
    public static async Task<Widget> ResolveReference([FromServices] WidgetRepository widgetData, [Id] string id, CancellationToken token)
    {
        // resolve the entity reference
        return await widgetData.GetWidgetByIdAsync(id, token);
    }
}
```

2. Via instance method, where the instance is the parsed representation.  Of course the instance would only have the representation properties populated.  The resolver could either populate the other properties and return itself, or extract the representation properties and return a new instance of the entity.  This might make the most sense to use if the representation returned child properties.

```cs
// configure federation key fields
[Key("id")]
public class Widget
{
    // configure fields
    [Id]
    public string Id { get; set; }
    public string Name { get; set; }

    // configure federation resolver
    [FederationResolver]
    public async Task<Widget> ResolveReference([FromServices] WidgetRepository widgetData, CancellationToken token)
    {
        // pull the id from the representation
        var id = Id;

        // resolve the entity reference
        return await widgetData.GetWidgetByIdAsync(id, token);
    }
}
```

## Findings

Unfortunately I realized that to properly implement variation 1 above requires a different design to the newly-added `IFederationResolver`.  Essentially, I had to move the code within `EntityResolver` that parsed representations into objects into `FederationResolverBase`, and allow `IFederationResolver` to dictate custom parsing code via the new `ParseRepresentation` method.

In addition, the coercion code needed to be duplicated from `ValidationContext.GetVariableValueAsync`.  I don't see a good alternative here, as the exceptions thrown by `ValidationContext.GetVariableValueAsync` are not applicable here.  This essentially makes 4 sets of coercion code within the codebase: (A) literal parsing (B) variable parsing (C) representation parsing into arguments (D) representation parsing into an object.  Three of the four are quite unique, but B and C could be potentially combined into some type of abstract class.

I've also added documentation within the migration notes for both code-first and type-first.  Of course, we should write a documentation page on all this.

Finally, I found that the `ResolveReference` extension methods were written incorrectly and missing some overloads to make them easier to use.  I would have realized this if I had some simple tests for the extension methods, which I should add.

## Option 3

Another option is to support `[FromSource]`.  I'd like to do this, so that it is possible for a person to pull the raw dictionary if desired, as shown below, or a custom representation class similar to `WidgetRepresentation` shown above.

```cs
// configure federation key fields
[Key("id")]
public class Widget
{
    // configure fields
    [Id]
    public string Id { get; set; }
    public string Name { get; set; }

    // configure federation resolver
    [FederationResolver]
    public static async Task<Widget> ResolveReference([FromServices] WidgetRepository widgetData, [FromSource] IDictionary<string, object?> args, CancellationToken token)
    {
        // pull the id from the representation
        var id = (string)args["id"];

        // resolve the entity reference
        return await widgetData.GetWidgetByIdAsync(id, token);
    }
}
```

This PR does not include 'option 3', but it is a possibility for v8.1, or some slight variation.  (Currently `[FromSource]` will throw an exception if the type does not match properly.)  Also, since the implementation details for the attribute are all private, we can refactor as desired.  It is likely enough that the two `Context` classes can all be unified into a single class while also supporting this 'option 3'.

## Todo

- [x] Add tests for code-first extension methods' usage
- [x] Add tests for type-first attribute usage
- [x] Add tests for type-first coercion
- [x] Add tests for code-first and type-first when multiple sets of key fields are defined

### CODE REVIEW

**Comment by gao-artur** [comment]
Not directly related to this PR, but I think the current implementation doesn't allow [multiple keys](https://www.apollographql.com/docs/federation/entities-advanced/#multiple-keys) on the entity. Adding support for it doesn't seem to affect the resolvers API, only the internal implementation. I mean, for static resolver, it should be possible just to define another resolver method:
```c#
[Key("id")]
[Key("name")]
public class Widget
{
    ...
    
    [FederationResolver]
    public static async Task<Widget> ResolveReferenceById([FromServices] WidgetRepository widgetData, [Id] string id, CancellationToken token)
    {
        ...
    }

    [FederationResolver]
    public static async Task<Widget> ResolveReferenceByName([FromServices] WidgetRepository widgetData, string name, CancellationToken token)
    {
        ...
    }
}
```

For an instance method, the user should be able to check the object properties to choose the correct resolver
```c#
[Key("id")]
[Key("name")]
public class Widget
{
    ...
    [FederationResolver]
    public async Task<Widget> ResolveReference([FromServices] WidgetRepository widgetData, CancellationToken token)
    {
        if (Id != null)
        {
             return await widgetData.GetWidgetByIdAsync(Id, token);
        }
        else if (Name != null)
        {
             return await widgetData.GetWidgetByNameAsync(Name, token);
        }
    }
}
```

Similarly, for code-first.

**Comment by gao-artur** [comment]
I checked the code again. I think I was wrong. Multiple keys are supported, but multiple resolvers are not.

**Comment by gao-artur** [comment]
I'm going on vacation until the 20th and will be far far away from my laptop, so won't be able to review your PRs

**Comment by Shane32** [comment]
> I'm going on vacation until the 20th and will be far far away from my laptop, so won't be able to review your PRs

That’s fine; no real rush here. Thanks for the heads-up!

**Comment by Shane32** [comment]
> I checked the code again. I think I was wrong. Multiple keys are supported, but multiple resolvers are not.

I've added documentation to the migration notes to explain this, and I'll make a note to write tests for this scenario.

**Comment by Shane32** [comment]
> for static resolver, it should be possible just to define another resolver method

True, that would be ideal, but it could be complex to implement.  Let's put that on a wish list for 8.1.  I think we can do it.

**Comment by Shane32** [comment]
> > for static resolver, it should be possible just to define another resolver method
> 
> True, that would be ideal, but it could be complex to implement. Let's put that on a wish list for 8.1. I think we can do it.

I added the feature now.

**Comment by Shane32** [comment]
@gao-artur Ready for review when you're back from your vacation.  Hope you're having a good time!

**Review by gao-artur — COMMENTED** [review]
Looks good!

File: src/GraphQL/Federation/Attributes/FederationResolverAttribute.FederationStaticResolver.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+            return true;
+        }
+
+        public object ParseRepresentation(IObjectGraphType graphType, IDictionary<string, object?> representation)
+        {
+            // Creates a dictionary of arguments for the field type based on the entity representation properties.
+            // The argument dictionary is returned as the parsed representation, to be used by the synthesized IResolveFieldContext.
+
+            if (_fieldType.Arguments == null || _fieldType.Arguments.Count <= 0)
```
[the remark this answers is not in the document]
Comment (Shane32) [reply src/GraphQL/Federation/Attributes/FederationResolverAttribute.FederationStaticResolver.cs:60]: I don't think so.
```suggestion
            if (_fieldType.Arguments == null || _fieldType.Arguments.Count == 0)
```

**Review by gao-artur — APPROVED** [review]

**Comment by Shane32** [comment]
@gao-artur I think there's basically just two PRs left.  The next one is adding `.AddFederation()` and related code to configure the schema.  And finally I'll add the test projects and stuff in a final PR.

**Pull request merged**